packages/overture-schema-core/src/overture/schema/core/scoping/scoped.py

```diff
@@ -0,0 +1,495 @@
+                    "from the start of the geometry, that the containing "
+                    f"{parent} applies to."
+                )
+            case Scope.HEADING:
+                return (
+                    "The heading, either forward or backward, that the "
+                    f"containing {parent} applies to."
+                )
+            case Scope.TEMPORAL:
+                return (
+                    "The recurring time span, in the OpenStreetMap opening "
+                    f"hours format, that the containing {parent} applies to. "
+                    "For the OSM opening hours specification, see "
+                    "https://wiki.openstreetmap.org/wiki/Key:opening_hours/specification."
+                )
+            case Scope.TRAVEL_MODE:
+                return (
+                    "A list of one or more travel modes, such as car, truck, "
+                    f"or foot, that the containing {parent} applies to."
+                )
+            case Scope.PURPOSE_OF_USE:
+                return (
+                    "A list of one or more usage purposes, such as delivery or "
+                    "arrival at final destination, that the containing "
+                    f"{parent} applies to."
+                )
+            case Scope.RECOGNIZED_STATUS:
+                return (
+                    "A list of one or more recognized status values, such as "
+                    f"employee or student, that the containing {parent} "
+                    "applies to."
+                )
+            case Scope.SIDE:
+                return (
+                    "The side, either left or right, that the containing "
+                    f"{parent} applies to."
+                )
+            case Scope.VEHICLE:
+                return (
+                    "A list of one or more vehicle parameters that limit the "
+                    f"vehicles the containing {parent} applies to."
+                )
+            case _:
+                raise self._unexpected_scope()
+
+    def _unexpected_scope(self) -> RuntimeError:
+        return RuntimeError(f"unexpected scope: {self}")
+
+    @staticmethod
+    def _top_level_scopes() -> tuple["Scope", ...]:
+        return (Scope.GEOMETRIC_POSITION, Scope.GEOMETRIC_RANGE, Scope.SIDE)
+
+    @staticmethod
+    def _when_scopes() -> tuple["Scope", ...]:
+        return (
+            Scope.HEADING,
+            Scope.TEMPORAL,
+            Scope.TRAVEL_MODE,
+            Scope.PURPOSE_OF_USE,
+            Scope.RECOGNIZED_STATUS,
+            Scope.VEHICLE,
+        )
+
+
+def scoped(
+    *optional: Scope,
+    required: Scope | Iterable[Scope] | None = None,
+) -> Callable:
+    """
+    Returns a decorator to decorate a Pydantic model class with one or more scoping attributes.
+
+    At least one `Scope` must be given either in `optional` or `required` or both. A scope may be
+    optional or required, but not both.
```

Comment:
How does one indicate that *a* scope must be provided without indicating which one?

Reply:
That's currently not possible.

I didn't think it was needed for any existing use case, but if I'm wrong let me know so I can fix it.

(Assuming we don't use that construct today: I can imagine it's something we'll need in the future. The best thing to do would be to drop `when` and just use `@require_any_of` to indicate this constraint. With `when` in the mix, it's messy.)